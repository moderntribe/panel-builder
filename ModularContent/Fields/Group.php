<?php


namespace ModularContent\Fields;

use ModularContent\Panel;
use ModularContent\AdminPreCache;

/**
 * Class Group
 *
 * A container for a group of fields. It wraps fields in the admin
 * in a fieldset to show logical groupings.
 *
 */
class Group extends Field {

	/** @var Field[] */
	protected $fields = [];

	protected $default = [];

	protected $layout = 'full';

	public function __construct( $args = [] ) {
		$this->defaults['layout'] = $this->layout;
		parent::__construct( $args );
	}

	/**
	 * @param Field $field
	 *
	 */
	public function add_field( Field $field ) {
		$this->fields[ $field->get_name() ] = $field;
	}

	/**
	 * @param $name
	 *
	 * @return Field|NULL
	 */
	public function get_field( $name ) {
		foreach ( $this->fields as $field ) {
			if ( $field->get_name() == $name || $this->get_name() . '.' . $field->get_name() == $name ) {
				return $field;
			}
		}

		return null;
	}

	/**
	 * Child fields should have the opportunity to set their own vars
	 *
	 * @param mixed $data
	 * @param Panel $panel
	 *
	 * @return array
	 */
	public function get_vars( $data, $panel ) {
		$vars = [];
		foreach ( $this->fields as $field ) {
			$name = $field->get_name();
			if ( isset( $data[ $name ] ) ) {
				$vars[ $name ] = $field->get_vars( $data[ $name ], $panel );
			}
		}

		$vars = apply_filters( 'panels_field_vars', $vars, $this, $panel );

		return $vars;
	}

	/**
	 * Child fields should have the opportunity to set their own vars for API
	 *
	 * @param mixed $data
	 * @param Panel $panel
	 *
	 * @return array
	 */
	public function get_vars_for_api( $data, $panel ) {
		$vars = [];
		foreach ( $this->fields as $field ) {
			$name = $field->get_name();
			if ( isset( $data[ $name ] ) ) {
				$vars[ $name ] = $field->get_vars_for_api( $data[ $name ], $panel );
			}
		}

		$vars = apply_filters( 'panels_field_vars_for_api', $vars, $data, $this, $panel );

		return $vars;
	}

	/**
	 * Add data relevant to this field to the precache
	 *
	 * @param mixed         $data
	 * @param AdminPreCache $cache
	 *
	 * @return void
	 */
	public function precache( $data, AdminPreCache $cache ) {
		foreach ( $this->fields as $field ) {
			$name = $field->get_name();
			if ( isset( $data[ $name ] ) ) {
				$field->precache( $data[ $name ], $cache );
			}
		}
	}

	public function get_blueprint() {
		$blueprint           = parent::get_blueprint();
		$blueprint['fields'] = [];
		foreach ( $this->fields as $field ) {
			$blueprint['fields'][] = $field->get_blueprint();
		}
		$blueprint['layout'] = $this->layout;

		return $blueprint;
	}

	/**
	 * Ensure that the submitted array is keyless
	 *
	 * @param array $data
	 *
	 * @return array
	 */
	public function prepare_data_for_save( $data ) {
		foreach ( $this->fields as $field ) {
			$name = $field->get_name();
			if ( ! isset( $data[ $name ] ) ) {
				$data[ $name ] = null;
			}
			$data[ $name ] = $field->prepare_data_for_save( $data[ $name ] );
		}

		return $data;
	}
} 