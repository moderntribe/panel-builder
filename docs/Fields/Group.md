## Group

A group of other Field Types.

### Parameters

||||
|---|---|---|
| `label`       | `string` | The __() translated label for this field       |
| `name`        | `string` | The unique name for this field                 |
| `description` | `string` | The __() translated description for this field |
| `layout` | `string` | The layout for this group, either `full` or `compact` |

### Example

```php
 $first_name = new Text( array(
   'label' => __('First Name'),
   'name' => 'first',
 ) );
 $last_name = new Text( array(
   'label' => __('Last Name'),
   'name' => 'last',
 ) );
 $group = new Group( array(
   'label' => __('Name'),
   'name' => 'name',
   'layout' => 'full',
 ) );
 $group->add_field( $first_name );
 $group->add_field( $last_name );
```

### Using data from a group in a template:

```php

 $first_name = get_panel_var( 'name.first' );
 $last_name = get_panel_var( 'name.last' );

 OR

 $name = get_panel_var( 'name' );
 $first_name = $name['first'];
 $last_name = $name['last'];

 OR

 $vars = get_panel_vars();
 $first_name = $vars['name']['first'];
 $last_name = $vars['name']['last'];
```