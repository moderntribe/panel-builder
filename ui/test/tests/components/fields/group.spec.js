/* global expect b:true */

import React from 'react';
import { shallow, mount } from 'enzyme';
import Group from 'components/fields/group';
import renderer from 'react-test-renderer';

describe('Group field', () => {
	it('renders correctly', () => {
		const tree = renderer
			.create(<Group />)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('renders as a <div>', () => {
		const wrapper = shallow(<Group />);
		expect(wrapper.type()).toEqual('div');
	});

	it('allows us to set props', () => {
		const wrapper = mount(<Group description="testing description" />);
		expect(wrapper.props().description).toEqual('testing description');
		wrapper.setProps({ description: 'Change props for test' });
		expect(wrapper.props().description).toEqual('Change props for test');
	});
});
