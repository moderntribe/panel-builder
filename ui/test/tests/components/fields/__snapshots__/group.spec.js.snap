exports[`Group field renders correctly 1`] = `
<div
  className="field panel-field group-field"
  data-group-active={false}>
  <div
    className="header panel-row-header"
    onClick={[Function]}>
    <h3>
      
    </h3>
    <i
      className="dashicons arrow panel-row-arrow dashicons-arrow-right-alt2" />
  </div>
</div>
`;
